airbyte-integrations/connectors/source-marketo/source_marketo/source.py

```diff
@@ -0,0 +1,530 @@
+
+    primary_key = "marketoGUID"
+    cursor_field = "activityDate"
+
+    def __init__(self, config: Mapping[str, Any]):
+        super().__init__(config, "activities")
+
+    @property
+    def stream_filter(self):
+        return {"activityTypeIds": [self.activity["id"]]}
+
+    def get_json_schema(self) -> Mapping[str, Any]:
+        properties = {
+            "marketoGUID": {"type": ["null", "string"]},
+            "leadId": {"type": ["null", "integer"]},
+            "activityDate": {"type": ["null", "string"], "format": "date-time"},
+            "activityTypeId": {"type": ["null", "integer"]},
+            "campaignId": {"type": ["null", "integer"]},
+            "attributes": {"type": ["null", "string"]},
+        }
+
+        if "primaryAttribute" in self.activity:
+            properties["primaryAttributeValue"] = {"type": ["null", "string"]}
+            properties["primaryAttributeName"] = {"type": ["null", "string"]}
+            properties["primaryAttributeValueId"] = {"type": ["null", "string"]}
+
+        if "attributes" in self.activity:
+            for attr in self.activity["attributes"]:
+                attr_name = clean_string(attr["name"])
+
+                if attr["dataType"] in ["datetime", "date"]:
+                    field_schema = {"type": "string", "format": "date-time"}
+                elif attr["dataType"] in ["integer", "percent", "score"]:
+                    field_schema = {"type": "integer"}
+                elif attr["dataType"] in ["float", "currency"]:
+                    field_schema = {"type": "number"}
+                elif attr["dataType"] == "boolean":
+                    field_schema = {"type": "boolean"}
+                elif attr["dataType"] in STRING_TYPES:
+                    field_schema = {"type": "string"}
+                elif attr["dataType"] in ["array"]:
+                    field_schema = {"type": "array", "items": {"type": ["integer", "number", "string", "null"]}}
+                else:
+                    field_schema = {"type": "string"}
+
+                field_schema["type"] = [field_schema["type"], "null"]
+
+                properties[attr_name] = field_schema
+
+        schema = {
+            "$schema": "http://json-schema.org/draft-07/schema#",
+            "type": ["null", "object"],
+            "additionalProperties": False,
+            "properties": properties,
+        }
+
+        return schema
+
+
+class ActivityTypes(MarketoStream):
+    """
+    Return list of all activity types.
+    API Docs: http://developers.marketo.com/rest-api/lead-database/activities/#describe
+    """
+
+    def path(self, stream_slice: Mapping[str, Any] = None, **kwargs) -> str:
+        return "/rest/v1/activities/types.json"
+
+
+class Programs(IncrementalMarketoStream):
+    """
+    Return list of all programs.
+    API Docs: http://developers.marketo.com/rest-api/assets/programs/#by_date_range
+    """
+
+    cursor_field = "updatedAt"
+    page_size = 200
+    offset = 0
+
```

Comment:
this is class variable. And it is updated by objects. It would be safer if it is initiated for every object separately.

Reply:
Updated.